java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToBQDstream.java

```diff
@@ -0,0 +1,200 @@
+
+    JavaSparkContext sparkContext = JavaSparkContext.fromSparkContext(spark.sparkContext());
+
+    JavaStreamingContext ssc = new JavaStreamingContext(sparkContext, new Duration(batchInterval));
+
+    JavaInputDStream<ConsumerRecord<Object, Object>> stream =
+        KafkaUtils.createDirectStream(
+            ssc,
+            LocationStrategies.PreferConsistent(),
+            ConsumerStrategies.Subscribe(topics, kafkaParams));
+
+    stream.foreachRDD(
+        (VoidFunction2<JavaRDD<ConsumerRecord<Object, Object>>, Time>)
+            (rdd, time) -> {
+              LOGGER.debug("Reading kafka data");
+
+              OffsetRange[] offsetRanges = ((HasOffsetRanges) rdd.rdd()).offsetRanges();
+
+              JavaRDD<Tuple2<String, String>> recordRdd =
+                  rdd.map(record -> new Tuple2(record.key(), record.value()));
+
+              Dataset<Row> rowDataset =
+                  spark
+                      .createDataset(
+                          recordRdd.rdd(), Encoders.tuple(Encoders.STRING(), Encoders.STRING()))
+                      .withColumnRenamed("_1", "key")
+                      .withColumnRenamed("_2", "value");
+
+              if (!rowDataset.isEmpty()) {
+
+                LOGGER.info("Saving data into BQ");
+
+                rowDataset
+                    .write()
+                    .mode(bqWriteMode)
+                    .format("bigquery")
+                    .option(
+                        KAFKA_BQ_SPARK_CONF_NAME_TABLE,
+                        projectId + "." + bigQueryDataset + "." + bigQueryTable)
+                    .option(KAFKA_BQ_SPARK_CONF_NAME_TEMP_GCS_BUCKET, tempGcsBucket)
+                    .option(KAFKA_BQ_SPARK_CONF_NAME_OUTPUT_HEADER, true)
+                    .save();
+
+                LOGGER.info("Saved to BQ");
+              }
+              ((CanCommitOffsets) stream.inputDStream()).commitAsync(offsetRanges);
+            });
+
+    ssc.start();
+    ssc.awaitTerminationOrTimeout(kafkaAwaitTerminationTimeout);
+  }
+
+  @Override
+  public void validateInput() {
+    if (StringUtils.isAllBlank(kafkaBootstrapServers)
+        || StringUtils.isAllBlank(kafkaTopic)
+        || StringUtils.isAllBlank(bigQueryDataset)
+        || StringUtils.isAllBlank(bigQueryTable)
+        || StringUtils.isAllBlank(projectId)
+        || StringUtils.isAllBlank(tempGcsBucket)) {
+      LOGGER.error(
+          "{},{},{},{},{},{} is required parameter. ",
+          PROJECT_ID_PROP,
+          KAFKA_BOOTSTRAP_SERVERS,
+          KAFKA_TOPIC,
+          KAFKA_BQ_DATASET,
+          KAFKA_BQ_TABLE,
+          KAFKA_BQ_TEMP_GCS_BUCKET);
+      throw new IllegalArgumentException(
+          "Required parameters for KafkaToBQDstream not passed. "
+              + "Set mandatory parameter for KafkaToBQDstream template "
+              + "in resources/conf/template.properties file.");
+    }
+
+    LOGGER.info(
+        "Starting Kafka to BQ spark job with following parameters:"
```

Comment:
`Kafka to BQ` -> `Kafka to BQ via DStream`

Reply:
done